PR review #1

jyoung90ie opened the pull request:
All trip changes

jyoung90ie committed:
Pages, layout and style created

jyoung90ie committed:
Initial deployment setup

jyoung90ie committed:
HTML templates setup for Flask

jyoung90ie committed:
Basic flask routing added

jyoung90ie committed:
added pymongo

jyoung90ie committed:
File renamed to show_trips.html

jyoung90ie committed:
MongoDB connection added

jyoung90ie committed:
environ vars and mongo conn updated

jyoung90ie committed:
changed env variable

jyoung90ie committed:
changed env variables

jyoung90ie committed:
updated env variables

jyoung90ie committed:
Merge branch 'trips' of https://git.heroku.com/jy-travelpal into trips

update branch

jyoung90ie committed:
update env variables

jyoung90ie committed:
Template for creating new trips

jyoung90ie committed:
Template for displaying all trips

jyoung90ie committed:
Base jinja2 template

jyoung90ie committed:
Template for viewing detailed trip info

jyoung90ie committed:
Templates for creating and updating trips

jyoung90ie committed:
Template for adding stops to a trip

jyoung90ie committed:
Template to test Flask pass-through routing

jyoung90ie committed:
Trips backend CRUD functionality added

jyoung90ie committed:
Updated breadcumb nav

jyoung90ie committed:
Updated error handling and added DB update functionality

jyoung90ie committed:
Tweak to show country name

jyoung90ie committed:
Updated to handle separate users

jyoung90ie committed:
User can now select 'all trips' or just their 'own trips'

jyoung90ie committed:
Browser validation and helper text added

jyoung90ie committed:
Detailed trip view now allows user to add and remove stops

jyoung90ie committed:
Navigation links updated

jyoung90ie committed:
UserID and Public flag uncommented for testing

jyoung90ie committed:
User feedback messages added

Flask flash messages have been implemented, so that each time the user
they receive feedback acknowledging the task. This helps provide a more
integrated user experience, and bridges the gap when redirecting to a
new page.

jyoung90ie committed:
Implement full CRUD functionality for 'trips' and 'stops'

Flask routing and templates created which enable the user to perform the
following tasks:
1) Trips: Add, Update, Remove
2) Stops: Add, Duplicate, Update, Remove

When a 'trip' is removed, all 'stops' associated with the trip are also
removed.

Flash messages have been added to provide user feedback throughout.

Collections have been stored in variables to streamline code.

A dummy function has been created to check for user permissions, this
will be implemented in the 'users' branch, following
the merging of the 'trips' branch to master.

jyoung90ie committed:
Update templates to respond to whether the user is logged in or not

This is preparatory work for
implementation of user functionality. The templates will only
display information that is relevant to the end-user.

For example, if the user is logged in and is the owner of the trip,
they will be shown extra options such as update and delete.

jyoung90ie committed:
Removes old user functions and updates navbar to respond to user login

This is preparatory work for implementation of full user login
functionality.

The navigation bar links have been updated to
display a set of links dependent on if the user is logged in or not.

The javascript fetch code has been removed as this will no longer be
used. The flask route was also removed from the app.py.

jyoung90ie committed:
Adds template for updating a 'stop'

New template created which is used to enable the user to update existing
stops. This is effectively the same as the stop_new.html template with
the addition of a pass-through stop query, which provides pre-populated
values for the form inputs.

This template is also used for the 'duplicate stop' functionality. A
stop is duplicated, then the user is redirected to this template to
perform any changes to the duplicated entry.

jyoung90ie committed:
Merge pull request #1 from jyoung90ie/trips

Trip functionality has been implemented. This enables end-users to add, remove, update, and duplicate entries, for 'trips' and 'stops'

jyoung90ie committed:
Adds user functionality

By building user functionality, this enables users to create their
own trips, which they have full control over - including whether
they wish to display the trip(s) publicly or not.

End users can now register an account, login and logout of the account.
This combined with previous updates will now provide a user specific
experience, i.e. only the user whom owns a 'trip' will have the ability
to change it.

The addition of checkUserPermission() enables efficient checking of user
permissions, for example, is the user logged in? does the user own
the trip in question? etc.

jyoung90ie committed:
Adds templates for registering new user and enabling user login

The templates combined with the changes to the app.py enable user
management.

Users login by inputting their username, this is checked to the database
if the user exists, a 'username' session variable is created which is
used throughout the templates to determine what the user should see.
The session variable stores the unique user id from the database.
This makes it easy to check whether the user is the owner of a trip.

User registration takes a few basic user details and checks the database
determine if this username already exists. If not, it is created.
If it already exists, the user is prompted to retry.

jyoung90ie committed:
Merge pull request #3 from jyoung90ie/users

Add user functionality

jyoung90ie committed:
Combine add and update into one template

To improve consistency and maintainability the add and update templates
have been merged into one template.

There is now one template for adding
and updating 'trips', and one for adding and updating 'stops'.

jyoung90ie committed:
Adds new field for trips to capture number of travelers

A new field has been added to capture the number of travelers per trip.
This will be used to scale the trip total costs. Additional text added
to input form to clarify what the costs represent.

Trips overview page is updated to include number of travelers and the
total cost is updated to multiply by the number of travelers.

jyoung90ie committed:
Styling tweaks

Breadcrumb styling updated and unused classes removed.

jyoung90ie committed:
Merge pull request #4 from jyoung90ie/combine-add-update

Combine add update

jyoung90ie committed:
Added new session variable

Display name has been pulled from the database and stored in a session
variable.

This is used for a welcome message when the user logs in to their
account.

jyoung90ie committed:
Classes updated

Removed all unused styles, as a result of html styling tweaks.

Added new class for welcome message.

jyoung90ie committed:
Adds welcome message to navbar and removes unnecessary whitespace

New welcome message added to the navbar if the user has logged in.

Added '-' modifiers to jinja2 templating code to remove whitespace
before and after injection of html.

jyoung90ie committed:
Updates the page layout to make use of Materialize cards

This update makes use of the extra real estate on larger screens,
making the design more responsive.

The layout is cleaner with all non-essential information hidden by
default. The additional information will be displayed upon request by
the user.

jyoung90ie committed:
Updates to provide a cleaner layout with responsive elements

Responsive design changes to make use of the additional real
estate of larger screens.

The overview section has been updated to provide the user with more
aggregate trip information.

The stops section has been updated to use a collapsible list. This
reduces the vertical space requirements and provides only the essential
information upfront, with further detail provided on demand. This makes
for a more positive user experience.

jyoung90ie committed:
Merge pull request #5 from jyoung90ie/layout-updates

Layout updates

jyoung90ie committed:
Adds docstrings

Docstrings have been added to all functions to provide a brief
description of their purpose.

jyoung90ie committed:
Updates templates to improve HTML validity

Unnecessary html and commented-out code have been removed.

Blank form actions have been removed and headers added within
each <section> so that the HTML is valid.

Other minor formatting and HTML tweaks also made to improve
website accessibility.

jyoung90ie committed:
Makes the breadcrumb trip name bold and consistent with other pages

jyoung90ie committed:
Removes redundant code and amends HTML and CSS styling

Unnecessary HTML elements have been removed and invalid HTML elements
are updated to make them valid.

Added materialize icons to collapsible list on detailed trip, all trips,
and my trips views, to let the user know they are interactive elements.

Jinja2 white space modifiers tweaked to make source code more readable.

jyoung90ie committed:
Updates card reveal layout

The card reveal content is updated to use a grid layout, to maintain the
clean user interface.

jyoung90ie committed:
Adds number of travelers

Template is updated to include the number of travelers.

jyoung90ie committed:
Isolates database connection and collection variables

Extracts database connection and collections variables into separate
file.

This enables the collection variables to be imported and used in
forms.py

jyoung90ie committed:
Updates database record from 'city/town' to 'city_town'

jyoung90ie committed:
Imports additional Flask form functionality and adds validation

Forms are now created using WTForms. This makes it easier to maintain
and enables robust validation of inputs.

jyoung90ie committed:
Updates templates to leverage WTForms functionality

Templates are now generated using WTForms. This enables
robust input validation and better error messages to isolate specific
invalid entries.

Going forward templates will now be easier to maintain.

jyoung90ie committed:
Updates app to import db and collection variables, and forms structure

MongoDB and collection variables have been moved to an external file to
enable them to be used in the forms.py file.

Trips, Stops, and, User functions have been updated to make use of
WTForms functionality. This results in more thorough validation and
data screening prior to inserting into the database.

This change also provides better data entry feedback, enabling users to
better self-correct data entry.

jyoung90ie committed:
Adds early wireframe document

jyoung90ie committed:
Trip detail view update to pull aggregate data

New MongoDB aggregate query added to add trip and stop level detail.
Trip level detail is produced by summing the totals across the trip
stops, this includes trip duration, total and per person costs, total
trip stops, total countries a trip visits, etc.

A similar breakdown is provided for stops leveraging trip dates and
trip travelers. This data is used to dynamically produce stop level
information, such as, total cost - using trip travelers with stop level
costs - and stop dates - using stop level duration and trip start date.

jyoung90ie committed:
Updates detailed trip view to only use one query

One aggregate query is used to pull the required information.

The query data is processed via a loop, with two data variables created
from this.

The first contains trip level detail which is used for the 'overview'
section.

The second contains stop detail, a list containing a dictionary for
each stop.

The two data variables are passed through to the template for display.

jyoung90ie committed:
Updates the cost formatting

Total cost now displays to two decimal places.

jyoung90ie committed:
Replaces placeholder text with data variables and tweaks formatting

The data being passed through from 'app.py' now contains the required
aggregate information. As a result all placeholder text (e.g.
_start_date_) has now been replaced with
variable names.

New total cost output added to trip overview for consistency across
views.

All variables which output a cost are now displayed to two decimal
places.

jyoung90ie committed:
Removes forms.py from ignore list

jyoung90ie committed:
Adds new classes and updates naming

New classes added for better layout on detailed trip view.

Added new 'error' class, this is used for form feedback messages.

Class 'trip-month' name changed so it is in-line with how the class is
used.

jyoung90ie committed:
Adds additional requirements used in python files

jyoung90ie committed:
Adds new function and updates trip listing aggregation query

getTripDuration() function added which returns the sum of the trip's
stop durations. This is used to calculate the trip end date.

Given the above, the trip collections field 'end_date' is no longer used
. It is instead populated with an empty value so that it may be used
for future updates.

The show_trips() view aggregation query has been updated to calculate
the end date for each trip, and to produce a listing of countries when
a trip is made up of more than one.

When adding or updating a trip stop, the trip end date field has been
replaced with trip projected end date. This projects the end date based
on the duration of the current stop summed with the duration of all
other stops for this specific trip.
It makes use of two hidden fields (total_trip_duration and
current_stop_duration) to calculate.

jyoung90ie committed:
Replaces trip end date with a projected end date based on user input

Two facilitate a projected end date, two hidden fields have been added.
They store the current stop duration and the trip total duration which
are used to calculate end date.

Given that there is no defined end date, it is no longer necessary to
run the validator check_duration - therefore removed from duration
field.

jyoung90ie committed:
Replaces trip end date with projected end date

Adds two hidden fields which store trip duration and current stop
duration (prior to being changed).

Trip end date field has been replaced with project end date field,
which is updated via the custom JavaScript setEndDate() function.

New JavaScript function created which updates the end date field when
the page loads and any time the user updates the duration field.
This also enforces a minimum duration of 1 (also checked on submission).

This provides the user with a more informed experience when updating
trips.

jyoung90ie committed:
Removes the end date field

End date is no longer hard coded, and is instead calculated based on
the total stop(s) duration.

jyoung90ie committed:
Updates trip_detailed view by adding total costs to trip and stop(s)

Variable trip_total_cost moved to outside loop to prevent double
counting.

New variables added, trip_total_cost_pp, stop_total_cost, and
stop_total_cost_pp, to provide better information to the user.

jyoung90ie committed:
Updates template to add trip and stop total costs

Additional total cost variables passed through from app.py, which have
been added to costs tables to provide better user information.

Trip overview variables grouped together in more meaningful way.

'Add stop' button moved to the left as it is the most likely first
action by the user when visiting this page.

jyoung90ie committed:
Updates template to use new variables passed through from app.py

Variable names have been updated to match the data passed through.

Trip listing layout updated to make better use of larger displays.

Data outputs have been split into rows so that there is no overlaps when
the data is larger than the column. Default margin between rows is
reduced via style.css

jyoung90ie committed:
Adds new classes for trip detailed and trip listing pages

Trip detailed: total column for tables updated to make it stand out from
non-total data.

Trip listing: added style to enable vertical scrolling of card data for
smaller devices. Also tweaked row margin spacing.

jyoung90ie committed:
Sets default duration value to 1

Duration form field default value set to 1 to prevent JavaScript onload
alert message.

jyoung90ie committed:
Updates queries to strip white space and format strings

Form string inputs into insert and update queries now use strip() to
remove white space.

All string inputs now have standardised formatting, as noted below:
- Trip Name, Stop Country and City/Town, and User 'Name', all formatted
using title case
- Username and Email formatted to lower case, for ease of comparison
later
- Currency formatted to uppercase

jyoung90ie committed:
Updates code to follow PEP conventions

- Added docstring
- Constant variables are now capitalised
- Reordered import sequence

Import sequence

jyoung90ie committed:
Updates code to follow PEP conventions

- Added docstring to module and functions
- Removed unused imports and reordered import sequence
- Constant variables are now capitalised
- Now using isinstance() instead of type()

jyoung90ie committed:
Updates code to follow PEP conventions

- Added docstring to module and functions
- Constant variables are now capitalised
- Function and variable names are now all snake_case
- Reordered import sequence
- Removed checkLogin from check_user_permssion as it was unused
- Added print statement for all exceptions
- Removed unnecessary variable declarations from trip_detailed view
- Updated the methodology for counting countries as previous method was
too simplistic and would double count if country appeared again
non-consecutively
- Removed unnecessary parentheses and else statements

jyoung90ie committed:
Removes default value for duration

This was removed as it was overriding value if user was updating
existing stop. Default value is now specified in app.py when form
invoked.

jyoung90ie committed:
Adds strong styling to trip name

To maintain consistency across pages, strong styling added to emphasis
the trip name.

jyoung90ie committed:
Updates README

Document updated with detailed headings and content.

jyoung90ie committed:
Fixes variable assignment

last_trip_start_date was being used before it was assigned. This update
fixes this.

ignatiusukwuoma commented on app.py:
```diff
@@ -0,0 +1,656 @@
+from datetime import datetime
+from flask_pymongo import PyMongo
+from dotenv import load_dotenv
+import os
```
A better ordering of imports will be import in this order
```
import Python's packages first eg. import os
import Framework/third party packages next eg. from flask_pymongo import PyMongo
import your own modules and packages last eg. from utils import get_names
```

ignatiusukwuoma commented on app.py:
```diff
@@ -0,0 +1,656 @@
+load_dotenv()
+
+app = Flask(__name__)
+# app.config['MONGO_DBNAME'] = 'travelPal'
```
remove uncommented code

ignatiusukwuoma commented on app.py:
```diff
@@ -0,0 +1,656 @@
+stops = mongo.db.stops
+
+
+def checkUserPermission(checkLogin=True, checkTripOwner=False,
```
The convention in python is to use snake case for function and variable names... 
```
check_user_permission instead of checkUserPermission
check_login instead of checkLogin
```

ignatiusukwuoma replied:
Let's have this function in a separate file called utils.py (to represent utility functions) then import into this file... so our app.py only contains routes

ignatiusukwuoma commented on app.py:
```diff
@@ -0,0 +1,656 @@
+
+#
+# trips functionality
+#
```
Use Docstrings instead https://www.python.org/dev/peps/pep-0257/

ignatiusukwuoma commented on templates/placeholder.html:
```diff
@@ -0,0 +1,9 @@
+{% block header %}{{ run_function }}{% endblock %}
+
+{% block content %}
+<div>ID: {{ trip_id }}</div>
```
Is this used?

jyoung90ie committed:
Updates the login process to transform username string to lowercase

The username string is transformed to lower case to compare to the
usernames stored in the database which are all lowercase.

This update takes place in the user_login view and the LoginForm
validation.

If this was not corrected a user would not be able to log in to their
account if they had any capitalisation in their input.

jyoung90ie committed:
Adds unit testing for user functionality

Using PyTest, a suite of unit testing functionality has been added to
ensure that user views are as expected, e.g. when logged in to a user
account you should not be able to register a new account.

jyoung90ie committed:
Renames 'db' to 'util' and updates references; fixes bug in trip_detail

Moves all functions from 'app', which are not route related, to
'util'. This makes the two files more distinct in their purpose and
makes it easier to find what you need, dependent on the nature of
what you're looking for.

Updates 'forms' to import from 'util' instead of 'db'.

The trip_detailed view within 'app' was not retaining the End Date
from the previous stop; added variable to fix this.

jyoung90ie committed:
Updates comments and docstrings, and adds additional error checking

Additional commentary added to provide more contextual rationale, and
existing commentary also updated to be more specific.

Docstrings have been updated to use " instead of ' for consistency.

Removed all print() statements to prepare for finalisation of document.

jyoung90ie committed:
Moves custom form validation to util and updates docstrings

Custom form validation has been moved to util, so that this file
specifically relates to form setup.

Docstrings have been updated to use " instead of ' for consistency.

jyoung90ie committed:
Adds form validation functions and updates docstrings

Form validation functions added so that all custom functionality is
contained within the util module.

Docstrings have been updated to use " instead of ' for consistency.

Removed all print() statements to prepare for finalisation of document.

jyoung90ie committed:
Removes duplicate entry of hidden fields

Using form.hidden_tag() displays all hidden fields. Calling specific
hidden fields following this resulted in duplicate rendering. Removed
individual rendering to stop this.

jyoung90ie committed:
Updates message when there are no trips to show

Message is now displayed as a header to give it more significance. The
message display is now dependent on whether the user is logged in or
not.

jyoung90ie committed:
Adds check_id function to test that URL id's are valid ObjectId

Additional validation function check_id added to ensure that the
trip_id and stop_id passed through the URL are valid.

This is needed to prevent users from manually modifying the URL which
will return an error.

jyoung90ie committed:
Adds 'Display publicly' information for the user

Updates the template to show if the user has opted to display their
trip to the public or not.

jyoung90ie committed:
Adds full application testing using PyTest

Unit Testing added to test the application routes, permissions, and
data entry.

The testing passes through valid and invalid data, testing each
independently to ensure that outcomes are being treated appropriately.

The application passes all the testing set out within the test_user.py
file.

jyoung90ie committed:
Updates the trip_detailed view for better error handling

Checks to determine if the trip_detailed query returns any results have
been updated.

Following unit testing, it was apparent that the checks to determine if
a trip exists or not were not working. The checks were taking
run too early in the process. They have been moved and updated to
correctly redirect the user when a trip does not exist.

Flash messages have had full stops added to the end for consistency.

jyoung90ie committed:
Adds additional sections and updates automated testing section

The automated testing section has been updated to add a high level
overview of the testing being conducted.

The deployment section has been split into three sub-sections, one for
local deployment, one for Heroku deployment, and the remaining details
the environment variables required.

New Credits section added to acknowledge media sourced from third-party.

jyoung90ie committed:
Adds additional commentary to the technologies section describing use

jyoung90ie committed:
Updates except statements and removes unnecessary permissions checks

Bare except statements updated to capture all exceptions.

Removed unnecessary checks to determine whether the user is logged in or
not. These are not required because the same functions already check if
the user owns the trip/stop - so by default will check if they are
logged in to an account.

jyoung90ie committed:
Updates tests to add valid and invalid data flag and refactors code

The tests for inputting data now all have a flag to indicate whether the
data is valid. This valid_entry flag will determine which unit test to
perform on an entry-by-entry basis. This provides a more complete
testing framework.

Code refactored to remove unnecessary functions which are now all passed
through the submit_form function. This makes the code easier to read
and maintain.

jyoung90ie committed:
Adds missing semi-colon to line 120

jyoung90ie committed:
Adds a brief paragraph describing the purpose of the application

jyoung90ie committed:
Adds additional UX design rationale and new sub-section under Testing

The UX Design section has been updated to provide further rationale for
the decisions made which resulted in the finished UX.

The Testing section has been updated with a new sub-section, Results.
This details the output from the Manual and Automated testing, which
were addressed.

jyoung90ie committed:
Updates the code to adhere to PEP8 guidance

Renamed id variable to id_value to be consistent with snake case.

Removed trailing white space and added a blank line at the end
consistent.

jyoung90ie committed:
Merge pull request #6 from jyoung90ie/development

Development
